cpp/ql/src/Likely Bugs/Conversion/LossyFunctionResultCast.ql

```diff
@@ -1,18 +1,65 @@
 /**
  * @name Lossy function result cast
  * @description Finds function calls whose result type is a floating point type, and which are casted to an integral type.
- *              Includes only expressions with implicit cast and excludes function calls to ceil, floor and round. {This is a gcc check; doesn't seem wildly useful.}
+ *              Includes only expressions with implicit cast and excludes function calls to ceil, floor and round.
  * @kind problem
  * @id cpp/lossy-function-result-cast
  * @problem.severity warning
+ * @precision medium
+ * @tags correctness
  */
+
 import cpp
 
+predicate whitelist(string fName) {
+  fName = "ceil" or
+  fName = "ceilf" or
+  fName = "ceill" or
+  fName = "floor" or
+  fName = "floorf" or
+  fName = "floorl" or
+  fName = "nearbyint" or
+  fName = "nearbyintf" or
+  fName = "nearbyintl" or
+  fName = "rint" or
+  fName = "rintf" or
+  fName = "rintl" or
+  fName = "round" or
+  fName = "roundf" or
+  fName = "roundl" or
+  fName = "trunc" or
+  fName = "truncf" or
+  fName = "truncl"
+}
+
+predicate whitelistPow(FunctionCall fc) {
+  (
+    fc.getTarget().getName() = "pow" or
+    fc.getTarget().getName() = "powf" or
+    fc.getTarget().getName() = "powl"
+  ) and exists(float value |
+    value = fc.getArgument(0).getValue().toFloat() and
+    (value.floor() - value).abs() < 0.001
+  )
+}
+
+predicate whiteListWrapped(FunctionCall fc) {
+  whitelist(fc.getTarget().getName()) or
+  whitelistPow(fc) or
+  exists(ReturnStmt rs |
+    rs.getEnclosingFunction() = fc.getTarget() and
+    whiteListWrapped(rs.getExpr())
+  ) or
+  exists(ReturnStmt rs, Variable v |
+    rs.getEnclosingFunction() = fc.getTarget() and
+    rs.getExpr().(VariableAccess).getTarget() = v and
+    whiteListWrapped(v.getAnAssignedValue())
+  )
```

Comment:
The wrapper detection is failing to detect [`vim_round`](https://github.com/vim/vim/blob/ebc8402063bbd5475ed348469172d4e3e260ae83/src/evalfunc.c#L10086), and this causes a FP on vim. I suggest using `DataFlow::localFlow`.

Reply:
Good thinking, done.